lib/pac55xx/usart.c

```diff
@@ -0,0 +1,207 @@
+@param[in] data_bits unsigned 8 bit. One of USART_DATABITS_5/6/7/8.
+@param[in] stop_bits unsigned 8 bit. One of USART_STOPBITS_1/1P5/2.
+@param[in] parity unsigned 8 bit. One of USART_PARITY_DISABLE/ODD/EVEN/FORCE1/FORCE0
+*/
+void usart_configure_lcr(uint32_t usart, uint8_t data_bits, uint8_t stop_bits,
+				uint8_t parity) {
+	USART_LCR(usart) = USART_LCR_WLS(data_bits)
+					| ((stop_bits==USART_STOPBITS_2) ? USART_LCR_SBS : 0)
+					| USART_LCR_PSELPEN(parity);
+}
+
+/** @brief Enable Break Control
+Enables break control bit that forces TX pin to logic low.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_break_enable(uint32_t usart) {
+	USART_LCR(usart) |= USART_LCR_BCON;
+}
+
+/** @brief Disable Break Control
+Disables break control bit that forces TX pin to logic low.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_break_disable(uint32_t usart) {
+	USART_LCR(usart) &= ~USART_LCR_BCON;
+}
+
+/** @brief Enable Enhanced Mode
+Enable enhanced mode to generate interrupts when FIFO thresholds in FCR are reached.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_enhanced_enable(uint32_t usart) {
+	USART_EFR(usart) = USART_EFR_ENMODE;
+}
+
+/** @brief Disable Enhanced Mode
+Disable enhanced mode to generate interrupts when FIFO thresholds in FCR are reached.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_enhanced_disable(uint32_t usart) {
+	USART_EFR(usart) &= ~USART_EFR_ENMODE;
+}
+
+/** @brief Enable FIFOs
+Enable both TX and RX FIFOs. This must be set before setting the trigger levels.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_fifo_enable(uint32_t usart) {
+	USART_FCR(usart) |= USART_FCR_FIFOEN;
+}
+
+/** @brief Disable FIFOs
+Disable both TX and RX FIFOs.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+*/
+void usart_fifo_disable(uint32_t usart) {
+	USART_FCR(usart) &= ~USART_FCR_FIFOEN;
+}
+
+/** @brief Configure FIFO Control Register
+This register sets the TX and RX FIFO interrupt trigger levels if the FIFOs are enabled.
+@param[in] usart unsigned 32 bit. USART block register address base @ref usart_reg_base
+@param[in] tx_depth unsigned 8 bit. One of USART_FIFO_TRIG_1/2/4/14CHAR.
+@param[in] rx_depth unsigned 8 bit. One of USART_FIFO_TRIG_1/2/4/14CHAR.
+@return true if the FIFOs are enabled and the set was successful. false if FIFOs are disabled.
+*/
+bool usart_configure_fcr(uint32_t usart, uint8_t tx_depth, uint8_t rx_depth) {
+	if (USART_FCR(usart) & USART_FCR_FIFOEN) {
+		USART_FCR(usart) = USART_FCR_TXTL(tx_depth) | USART_FCR_RXTL(rx_depth)
+						| USART_FCR_FIFOEN;
+		return true;
+	}
+	/* do nothing if FIFOs aren't enabled */
```

Comment:
why not just have the function _do it_
```
USART_FCR |= FIFOEN | TXDEPTH | RXDEPTH;
```
What's the advantage of this? who would ever want to look at the return value here?

Reply:
The FIFOEN bit must be set prior to calling this function.  It has to be done in 2 separate writes to the register.  I wanted some way to indicate "you're doing it wrong" if `usart_fifo_enable()` wasn't called first.